### PR TITLE
feat: replace env/org hrids configuration by ids

### PR DESCRIPTION
**Pull request by guillaumelamirand**
## Issue

https://gravitee.atlassian.net/browse/ARCHI-386

## Description

Introduce a new way to configure environment id and organization id on the gateway side using directly environmentIds.

A new configuration `useLegacyEnvironmentHrids` could be set to configure how environments id are resolved.
The old configuration is still supported and will be the default behavior.
If `useLegacyEnvironmentHrids`is set to false, so only environments ID are supported with uuid only (or DEFAULT)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java [hunk 1]
```diff
@@ -65,35 +66,80 @@ public class GatewayNodeMetadataResolver implements NodeMetadataResolver {
...
+                    return Optional.<Environment>empty();
+                }
+            })
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .collect(Collectors.toSet());
+    }
+
+    private static boolean validateEnvironmentId(final String id) {
+        return id.equals("DEFAULT") || valideUUID(id);
```
Comment (benoitgravitee) [inline gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java]: maybe use `Objects.equals()` for more safety

File: gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java [hunk 2]
```diff
@@ -65,35 +66,81 @@ public class GatewayNodeMetadataResolver implements NodeMetadataResolver {
...
+    }
+
+    private static boolean valideUUID(final String id) {
+        try {
+            UUID.fromString(id);
+            return true;
+        } catch (Exception e) {
```
Comment (benoitgravitee) [inline gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java]: IllegalArgumentException would be a better choice so you don't other exceptions, unlikely but still...

**Review by benoitgravitee — APPROVED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java [hunk 3]
```diff
@@ -65,35 +67,81 @@ public class GatewayNodeMetadataResolver implements NodeMetadataResolver {
...
+            .map(id -> {
+                try {
+                    return environmentRepository.findById(id);
+                } catch (TechnicalException e) {
+                    log.warn("Unable to load environment from id {}, ignore it.", id, e);
+                    return Optional.<Environment>empty();
+                }
+            })
+            .filter(Optional::isPresent)
+            .map(Optional::get)
```
Comment (michel-barret) [inline gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java]: Can be simplified

```suggestion
            .flatMap(id -> {
                try {
                    return environmentRepository.findById(id).stream();
                } catch (TechnicalException e) {
                    log.warn("Unable to load environment from id {}, ignore it.", id, e);
                    return Stream.<Environment>empty();
                }
            })
```

Comment (guillaumelamirand) [reply gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java]: Honestly I don't feel it is more simple.

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/repository/synchronizer/node/NodeMetadataSynchronizer.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    NodeMetadataDeployer nodeMetadataDeployer = deployerFactory.createNodeMetadataDeployer();
+                    return upstream
+                        .subscribeOn(Schedulers.from(syncDeployerExecutor))
+                        .flatMap(deployable -> deploy(nodeMetadataDeployer, deployable));
+                })
+                .ignoreElement()
+                .doOnSubscribe(disposable -> launchTime.set(Instant.now().toEpochMilli()))
+                .doOnComplete(() -> {
+                    String logMsg = String.format("Node metadata synchronized in %sms", (System.currentTimeMillis() - launchTime.get()));
+                    log.info(logMsg);
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/repository/synchronizer/node/NodeMetadataSynchronizer.java]: You can directly format the log and avoid String.format

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/repository/synchronizer/node/NodeMetadataSynchronizer.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    private static Maybe<NodeMetadataDeployable> deploy(
+        final NodeMetadataDeployer nodeMetadataDeployer,
+        final NodeMetadataDeployable deployable
+    ) {
+        return nodeMetadataDeployer
+            .deploy(deployable)
+            .andThen(nodeMetadataDeployer.doAfterDeployment(deployable))
+            .andThen(Maybe.just(deployable))
+            .onErrorResumeNext(throwable -> {
+                log.error(throwable.getMessage(), throwable);
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/repository/synchronizer/node/NodeMetadataSynchronizer.java]: The message is already part of the throwable. Maybe `log.error("An error occurred during node metadata deployment", e);` would be better.

File: gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java [hunk 6]
```diff
@@ -65,35 +66,81 @@ public class GatewayNodeMetadataResolver implements NodeMetadataResolver {
...
+            .collect(Collectors.toSet());
+    }
+
+    private static boolean validateEnvironmentId(final String id) {
+        return Objects.equals(id, "DEFAULT") || valideUUID(id);
+    }
+
+    private static boolean valideUUID(final String id) {
```
Comment (jhaeyaert) [inline gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java]: Typo valideUUID -> validateUUID

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/InstanceResource.java [hunk 7]
```diff
@@ -81,14 +82,13 @@ public MonitoringResource getMonitoringResource() {
         return resourceContext.getResource(MonitoringResource.class);
     }
 
-    private boolean isInstanceAccessibleByOrga(List<String> organizationsHrids, String currentOrganization) {
-        if (organizationsHrids == null || organizationsHrids.isEmpty()) {
+    private boolean isInstanceAccessibleByOrga(Set<String> environmentIds, String currentOrganization) {
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/InstanceResource.java]: i believe isInstanceAccessibleByEnv and isInstanceAccessibleByOrga could now be merged into 1 single isInstanceAccessible

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/EnvironmentServiceImpl.java [hunk 8]
```diff
@@ -132,6 +132,22 @@ public EnvironmentEntity findByOrgAndIdOrHrid(final String organizationId, Strin
...
+        try {
+            LOGGER.debug("Find organization id from environment ids");
+
+            return environmentRepository
+                .findOrganizationIdsByEnvironments(environmentIds)
+                .stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+        } catch (TechnicalException ex) {
+            LOGGER.error("An error occurs while trying to find organization ids from environment ids", ex);
```
Comment (jhaeyaert) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/EnvironmentServiceImpl.java]: The log can be removed as the exception is propagated.

**Review by jhaeyaert — CHANGES_REQUESTED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/InstanceResource.java (hunk 7 above)
Comment (guillaumelamirand) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/InstanceResource.java]: The first checks the envs according to the env in the context, and the second checks the orgas of the envs according to the orga in the context so we still first validate about env then orga. Can you elaborate?

Comment (jhaeyaert) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/InstanceResource.java]: I initially thought it was possible to have 1 single method, but I dug a bit and believe we can do it another way. Let's discuss it.

**Review by jhaeyaert — APPROVED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java (hunk 3 above)
Comment (guillaumelamirand) [reply gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/src/main/java/io/gravitee/gateway/standalone/node/GatewayNodeMetadataResolver.java]: Code as been removed anyway. 👍

**Pull request merged**